### ISSUE TITLE
nftables kube-proxy TODO

### ISSUE THREAD

**Issue by danwinship** [kubernetes/kubernetes#122572]
#### For 1.30:
- [X] add periodic and presubmit e2e tests (https://github.com/kubernetes/test-infra/pull/31525)
- [ ] add performance job
  - Discussion on the KEP PR: https://github.com/kubernetes/enhancements/pull/3824#discussion_r1107584724, https://github.com/kubernetes/enhancements/pull/3824#discussion_r1328756780
- [ ] ensure unit test parity with iptables. (I think all that's left here is adding unit tests for the packet tracing code in `helpers_test.go`.)
- [ ] decide on required kernel/`nft` versions
  - the first half of #122296 has discussion of this before it turned into something else
  - actually, we've pretty much decided on kernel 5.6, though we will probably implement this via an `nft --check`-based check rather than via a kernel version number check. (Note that current knftables master now has a separate [`nft.Check()`](https://github.com/danwinship/knftables/commit/63b6682f2040bbc78bb6ad5c3bc164c16971fa7c) method to facilitate this.)
  - need to figure out corresponding minimum client version, and make sure our startup version check catches both "kernel too old" and "`/sbin/nft` too old".
  - should document the eventual requirement [here](https://kubernetes.io/docs/reference/networking/virtual-ips/#proxy-mode-nftables)
  - kube-proxy should error out at startup with a clear message if you have a too-old kernel or `nft`
- [ ] add some useful metrics
  - we don't know exactly what these will be: [discussion in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#monitoring-requirements)
  - clean up existing metrics stuff a bit at the same time? (e.g., https://github.com/danwinship/kubernetes/commits/proxy-metrics/)
- [ ] change `--nodeport-addresses` behavior to default to "primary node IP(s) only" rather than "all node IPs".
  - [discussion in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#the-iptables-mode-of-kube-proxy-has-grown-crufty)
  - Discussion in the KEP PR: https://github.com/kubernetes/enhancements/pull/3824#pullrequestreview-1284130849, https://github.com/kubernetes/enhancements/pull/3824#issuecomment-1418010527, https://github.com/kubernetes/enhancements/pull/3824#discussion_r1334533161
- [ ] decide whether to change NodePorts vs LoadBalancerSourceRanges behavior, and if so, do it
  - [discussion in iptables proxier source](https://github.com/kubernetes/kubernetes/blob/v1.29.0/pkg/proxy/iptables/proxier.go#L1167-L1169)
  - [discussion in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#the-iptables-mode-of-kube-proxy-has-grown-crufty)
- [ ] decide if we are going to change anything about session affinity
  - [Discussion in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#session-affinity). I'm not sure if anyone has suggested any particular benefits of the ipvs-like behavior vs the iptables-like behavior.
  - other discussion about changes to affinity behavior: #112806
  - Resolve the `UNRESOLVED` section of the KEP after a decision is made.
- [ ] `reject` connections on invalid ports of service IPs
  - [Discussion in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#behavior-of-service-ips)
  - kube-proxy could watch [`ServiceCIDR` objects](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/1880-multiple-service-cidrs/README.md) to learn the full service CIDR(s) and reject connections on _all_ service IPs, not just currently-in-use-ones.
  - Resolve the `UNRESOLVED` section of the KEP after this is implemented.
- [ ] decide if we are using comments well in the ruleset
  - We don't need many of the comments the iptables ruleset uses because our chain names are self-documenting.
  - In general I tried to put comments on objects (chains, sets, maps) rather than on individual rules, but there is some inconsistency. (But also, note that you won't see the chain/set/map comments with older kernels.)
  - Individual rules and set/map elements should have comments in some cases. E.g., [elements of `@no-endpoint-services` and `@no-endpoint-nodeports` have comments](https://github.com/kubernetes/kubernetes/blob/v1.29.0/pkg/proxy/nftables/proxier_test.go#L575-L577) so that you can see which IP/port goes with which service. But [elements of `@service-ips` and `@service-nodeports` don't](https://github.com/kubernetes/kubernetes/blob/v1.29.0/pkg/proxy/nftables/proxier_test.go#L570-L573) because you can already figure that out from the names of the chains that they jump to.
- [ ] move `danwinship/knftables` to `kubernetes-sigs/knftables` (and eventually declare a v0.1.0 API)

#### Things that depend on us having a perf job (and good metrics) first:
- [ ] add iptables-style partial syncing
- [ ] optimize memory allocation:
  - reuse the buffers used for building the `nft -f -` input, [like iptables does](https://github.com/kubernetes/kubernetes/blob/v1.29.0/pkg/proxy/iptables/proxier.go#L185-L192). This could be done in a few ways. (Have `Proxier` keep buffers around like iptables does and pass a buffer to `nft.Run()`; have `knftables.realNFTables` keep buffers around itself and pass them to the `Transaction`; have `knftables.realNFTables` keep `Transaction`s around and reuse them (with the transaction storing its buffer); ...)
  - maybe reuse the `knftables.Transaction.operation` arrays somehow?
  - consider optimizing `knftables.Rule` generation wrt `knftables.Concat`. Some discussion [here](https://github.com/kubernetes/kubernetes/pull/121046#discussion_r1367736334).
- [ ] rewrite masquerading to use an nftables set rather than using the mark, [as discussed in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#masquerading)
- [ ] try out the alternative hairpin rule [suggested in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#future-improvements).

#### Additional metrics for iptables mode to help users figure out if they'd have trouble migrating:
- [ ] Metric to indicate if you are using localhost nodeports. This could be done by making sure there is always a separate rule to catch 127.0.0.1:nodeport rather than letting it be caught by a generic rule, and then checking the counters on those rules and exporting some metric with the result.
- [ ] More generally, metric to indicate if you are using non-primary-node-IP nodeports. (Maybe the metric could just indicate every IP/interface that you're making use of nodeports on?)
- [ ] Metric to indicate if you are relying on NodePorts not being subject to LoadBalancerSourceRanges. (Similar to the above, this would require changing rules around so that there's some rule whose counter will indicate when you're relying on the current behavior). (Possibly an obscure enough edge case that this isn't worth checking, since it might be messy to check?)
- [ ] Possibly other metrics if we change other behavior

#### For 1.31/beta:
- [ ] Generally better documentation (mostly in `content/en/docs/reference/networking/virtual-ips.md` in k/website)
  - Document all [changes from iptables](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#changes-from-the-iptables-kube-proxy-backend), including anything above not mentioned in the KEP.
  - Document different interaction with firewalls. ([Discussion in KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#forcing-traffic-on-healthchecknodeports-to-be-accepted))
- [ ] periodic e2e test of migration/rollback between iptables/ipvs and nftables
- [ ] consider the suggestion of [making it possible to run multiple instances of kube-proxy](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#multiple-instances-of-kube-proxy)
  - Discussion in the KEP PR: https://github.com/kubernetes/enhancements/pull/3824#issuecomment-1574741149, https://github.com/kubernetes/enhancements/pull/3824#issuecomment-1576820872
  - Resolve the `UNRESOLVED` section of the KEP after a decision is made.

#### For GA:
- [ ] Document the "API guarantees" for components interoperating with nftables kube-proxy.
  - [Discussion in KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#defining-an-api-for-integration-with-admindebugthird-party-rules)
  - Resolve the `UNRESOLVED` section of the KEP when this is implemented.

/sig network
/priority important-soon
/triage accepted
cc @aojea @uablrek @aroradaman @tnqn

**Comment by uablrek** [comment]
I can easily switch kernels in my env (and "nft" for that matter), so I can check that. But what to check, and how precise?

Is it enough to verify basic traffic, or is a full K8s e2e required?

Even if I can switch kernels, download and building them is a tedious task (especially if there are gcc version problems), so the fewer the better.

**Comment by tnqn** [comment]
> I can run on linux-5.6.10 without problems, but on linux-5.6 and linux-5.6.4 I get errors:

@uablrek For v1.29.0, the minimal kernel version is linux-5.6.9 according to previous investigation: https://github.com/kubernetes/kubernetes/pull/122296#issuecomment-1864347600
> Yes, I managed to run it with 5.6 but only after 5.6.9. It's because nftables CLI "Set NFT_SET_CONCAT flag for sets with concatenated ranges" since [0.9.5](https://www.netfilter.org/projects/nftables/files/changes-nftables-0.9.5.txt), and the flag was added to [kernel 5.7](https://github.com/torvalds/linux/commit/ef516e8625ddea90b3a0313f3a0b0baa83db7ac2) and [backported to 5.6.9](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-5.6.y&id=2648315067b0700c862bc2cbf4338aeccd385698).

After https://github.com/kubernetes/kubernetes/pull/122296, the minimal version may be relaxed a lot. I guess it's [4.1](https://wiki.nftables.org/wiki-nftables/index.php/List_of_updates_since_Linux_kernel_3.13#4.1) but haven't tested it on that version yet.

**Comment by danwinship** [comment]
> And, is it OK to discuss in this PR?

I assume this PR is going to get a lot of discussion about a lot of things, and that's fine, but if you're planning to grab one of the items anyway, maybe file a separate issue for it (or an initial/placeholder PR) and move discussion there?

> Is it enough to verify basic traffic, or is a full K8s e2e required?

We need to verify that every kind of rule the nftables Proxier can generate works correctly. So somewhere in between; full e2e would do the trick, but in theory you could do something much simpler.

One starting point would be the `expected` ruleset from `TestOverallNFTablesRules` in `pkg/proxy/nftables/proxier_test.go`. Unfortunately, it looks like it's currently not in a loadable order. (e.g., it adds a `jump endpoints-check` rule before adding the `endpoints-check` chain.) You can hack around that by first running just the `add table` and `add chain` lines, and then after that re-run the entire transaction. (Probably we should fix the value of `expected` so that it is usable directly...) But also, this was mostly just a port of the corresponding iptables proxy test, and I don't know for sure that it actually includes a copy of every possible rule type. (Though it would obviously be good if it did.) So, (1) figure out what additional service types are needed and add them, (2) figure out if there are any rule types that can't be generated here because they'd require different global config (e.g. not having a local detector) and add comments about that, (3) re-order the rules into a usable order (while still trying to preserve the grouping-by-service as much as possible for debuggability), (4) test against kernels.

**Comment by danwinship** [comment]
> After #122296, the minimal version may be relaxed a lot. I guess it's [4.1](https://wiki.nftables.org/wiki-nftables/index.php/List_of_updates_since_Linux_kernel_3.13#4.1) but haven't tested it on that version yet.

Currently we use `th` expressions, which were added in 5.3, there's a workaround for that if needed. Other than that though, there are also bugfixes to worry about; the list of updates in the wiki includes really major ones, but I worry that the further back we go, the more minor bugs creep in that could make things not work. (Especially, if we wanted to support earlier than 4.18 we'd have to figure out what "nftables NAT is no longer incompatible with iptables NAT" means.) So I don't think we want to go any earlier than 5.4, which I believe is the oldest kernel in use in any LTS distro that's still in a "fully supported" phase.

**Comment by danwinship** [comment]
> Any idea on what to check with "nft --check"?

So, per [the wiki](https://wiki.nftables.org/wiki-nftables/index.php/List_of_updates_since_Linux_kernel_3.13), if we want to require 5.4, I guess you could try a rule with "`meta time`"... though it seems a little weird to have the test be based on something we aren't actually using, so maybe `th` would be better?

**Comment by danwinship** [comment]
Also, I haven't investigated the `nft` version situation at all, and there may be constraints there that make us need a newer `nft` than "whatever version added support for the oldest kernel feature we're using". Especially, I'm not sure exactly when `--json` support was added, and we definitely need that.

**Comment by tnqn** [comment]
> Currently we use `th` expressions, which were added in 5.3

It seems `th` expressions is only an alias and the change is needed in CLI only: [proto: add pseudo th protocol to match d/sport in generic way](https://git.netfilter.org/nftables/commit/?h=v0.9.2&id=a43a696443a150f448d182fe708a43d7279de715). See `As "th" is an alias for the raw expression, no dependency is generated`. And kernel [5.3](https://kernelnewbies.org/Linux_5.3) doesn't mention related change in kernel at all. I successfully ran nftables mode and saw a lot of `th` expression with kernel 4.15 and nftables v1.0.6.

> Especially, if we wanted to support earlier than 4.18 we'd have to figure out what "nftables NAT is no longer incompatible with iptables NAT" means.

It means before 4.18 iptables NAT rule will not take effect when nftables NAT rule is in use. [The page](https://wiki.nftables.org/wiki-nftables/index.php/Performing_Network_Address_Translation_(NAT)#Incompatibilities) describes it. And I have confirmed on kernel 4.15, after nftables rules are installed, the iptables nat rules are no longer matched, causing problems when Pods try to access external network via iptables MASQUERADE. So if a CNI plugin relies on iptables NAT table, it's not going to work with nftables mode before 4.18.

**Comment by danwinship** [comment]
> See `As "th" is an alias for the raw expression, no dependency is generated`.

"dependency" here is nftables-maintainer-speak. It's warning you that `th` doesn't imply `{tcp,udp,sctp}` so if you're careless with your rules, a `th` expression might end up matching bits inside an ICMP packet or something.

(I got the "5.3" requirement from [here](https://wiki.nftables.org/wiki-nftables/index.php/Matching_packet_headers#Matching_UDP.2FTCP_headers_in_the_same_rule) but I agree, I can't see any commits between 5.2 and 5.3 that seem relevant.)

**Comment by uablrek** [comment]
I have tested some versions of `nft` ([archives](https://netfilter.org/projects/nftables/files/)):
```
nftables v0.9.2 (Scram)                  2019-08-19
nftables v1.0.0 (Fearless Fosdick #2)    2021-08-19
```
Both works with K8s e2e with FOCUS/SKIP as in https://github.com/kubernetes/kubernetes/issues/122572#issuecomment-1876311340.

Can we settle for nft $\ge$ v0.9.2?

**Comment by danwinship** [comment]
> Can we settle for nft ≥ v0.9.2?

Yes. I think we'll officially declare that we support kernel 5.4+ and nftables command-line 0.9.2+, and we can use a `th` expression to check that. (Which means really we'll support kernel 5.3, but we won't advertise it that way.)

FTR (and maybe worth documenting in a comment wherever we put the check):
- If we decide we need concatenations-with-ranges support again later, that would imply 5.6/0.9.4.
- If we want to use `typeof` in sets/maps, that would require at least 0.9.4 (and maybe later depending on exactly how we're using it; we'd have to re-check the release notes).
- Support for comments on tables/chains/sets/maps/etc (anything but rules and elements) didn't come until 5.10/0.9.7, but knftables itself already does a `--check` for that at construct time, and just drops those comments if they're not supported.

**Comment by aroradaman** [comment]
For perf job do we need something similar to [ci-kubernetes-e2e-gce-scale-performance](https://github.com/kubernetes/test-infra/blob/dd8dfec9b2dd99999a1187fe861c90a2694556d3/config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml#L62C9-L62C48) configured for NFTables with lower cadence and nodes?

**Comment by danwinship** [comment]
I'm not sure exactly what we want for perf... @aojea may have thoughts on that, and sig-scalability may have more information on what sort of big tests we're _able_ to run. (I assume we can't spin up a 2000 node cluster just because we're bored...)

Maybe open an issue to start the discussion on that?

(Antonio also said he was going to start thinking about metrics that the perf job could be measuring.)

**Comment by nayihz** [comment]
> * change `--nodeport-addresses` behavior to default to "primary node IP(s) only" rather than "all node IPs".

Do we need to implement it **only** in nftables mode or in all modes(both in iptables and ipvs)?

And by the way, I think we should create a new issue for every task. This issue is not a good place to discuss details.

**Comment by npinaeva** [comment]
Will try to tackle `ensure unit test parity with iptables.` and `change --nodeport-addresses behavior` (@nayihz if you wanted to work on the latter, lmk and I will pick something else)

**Comment by nayihz** [comment]
> (@nayihz if you wanted to work on the latter, lmk and I will pick something else)

Yes, I want to work on it after we confirm some details.

**Comment by aroradaman** [comment]
picking up `reject connections on invalid ports of service IPs`.

**Comment by danwinship** [comment]
> Do we need to implement it **only** in nftables mode or in all modes(both in iptables and ipvs)?

We're only changing behavior of nftables mode. All of the todo items here are for nftables mode only, except for the "Additional metrics for iptables mode" section.

> And by the way, I think we should create new issue to track every task. This issue is not a good place to discuss details.

Yes, people should comment here if they want to claim an item, but it's best to split out discussion. You can either open an issue for the specific item, or else just file a Draft PR with some initial work and we can discuss there. (And I'll update the list at the top to link to other issues/PRs as people file them.)

**Comment by uablrek** [comment]
I will check [making it possible to run multiple instances of kube-proxy](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md#multiple-instances-of-kube-proxy). Not the most urgent, but fun :smile: 

It can be tested as a PoC with Multus, independent of [K8s multi-network](https://github.com/kubernetes/enhancements/pull/3700), but will be really useful with it. Load balancing is not yet considered in `K8s multi-network`, but I think is must at some point.

[Kpng](https://github.com/kubernetes-sigs/kpng) is already supporting multiple instances, so this is not impossible at all.

**Comment by uablrek** [comment]
### An update on `nft` versions

On KinD nft v0.9.8 segv's. On a KinD node:
```
apt update
apt install -y nftables
nft list ruleset
Segmentation fault (core dumped)
```
Stack trace:
```
Core was generated by `nft list ruleset'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007f3691345743 in set_make_key (attr=attr@entry=0x558676fe2f1c)
    at netlink.c:767
#2  0x00007f3691347bba in netlink_delinearize_set (
    ctx=ctx@entry=0x7ffdc4c346a0, nls=0x558676fed450) at netlink.c:841
#3  0x00007f3691348253 in list_set_cb (nls=<optimized out>, arg=0x7ffdc4c346a0)
    at netlink.c:978
#4  0x00007f36910d91d5 in nftnl_set_list_foreach ()
   from /usr/lib/x86_64-linux-gnu/libnftnl.so.11
#5  0x00007f36913482b8 in netlink_list_sets (ctx=ctx@entry=0x7ffdc4c346a0, 
    h=h@entry=0x558676fe1410) at netlink.c:999
#6  0x00007f369132e67d in cache_init_objects (flags=536871039, 
    ctx=0x7ffdc4c346a0) at rule.c:170
#7  cache_init (flags=536871039, ctx=0x7ffdc4c346a0) at rule.c:235
#8  cache_update (nft=nft@entry=0x558676fd72a0, flags=<optimized out>, 
    msgs=msgs@entry=0x7ffdc4c35660) at rule.c:291
#9  0x00007f369135cf12 in nft_evaluate (nft=nft@entry=0x558676fd72a0, 
    msgs=msgs@entry=0x7ffdc4c35660, cmds=cmds@entry=0x7ffdc4c35670)
    at libnftables.c:420
#10 0x00007f369135d8eb in nft_run_cmd_from_buffer (nft=0x558676fd72a0, 
    buf=0x558676fd7b30 "list ruleset") at libnftables.c:465
#11 0x0000558676fa58c6 in ?? ()
#12 0x00007f3691124d0a in __libc_start_main (main=0x558676fa5290, argc=3, 
    argv=0x7ffdc4c35818, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7ffdc4c35808)
    at ../csu/libc-start.c:308
#13 0x0000558676fa5a8a in ?? ()
```
For v1.0.3+ it seems to be:
```
nftables v0.9.2 (Scram)
OK
nftables v0.9.3 (Topsy)
OK
nftables v0.9.4 (Jive at Five)
Segmentation fault
nftables v0.9.5 (Capital Idea)
Segmentation fault
nftables v0.9.6 (Capital Idea #2)
Segmentation fault
nftables v0.9.7 (Anyface)
Segmentation fault
nftables v0.9.8 (E.D.S.)
Segmentation fault
nftables v0.9.9 (Prudence Pimpleton)
OK
nftables v1.0.0 (Fearless Fosdick #2)
OK
nftables v1.0.1 (Fearless Fosdick #3)
OK
nftables v1.0.2 (Lester Gooch)
OK
nftables v1.0.3 (Lester Gooch #2)
OK
nftables v1.0.4 (Lester Gooch #3)
OK
nftables v1.0.5 (Lester Gooch #4)
OK
nftables v1.0.6 (Lester Gooch #5)
OK
nftables v1.0.7 (Old Doc Yak)
OK
nftables v1.0.8 (Old Doc Yak #2)
OK
nftables v1.0.9 (Old Doc Yak #3)
OK
```

**Comment by aroradaman** [comment]
`registry.k8s.io/kube-proxy:v1.29.0` was built with `v1.0.6 (Lester Gooch #5)`

**Comment by danwinship** [comment]
moving versioning discussion to #122743

**Comment by uablrek** [comment]
Multiple instances of kube-proxy in https://github.com/kubernetes/kubernetes/pull/122814

**Comment by aroradaman** [comment]
@danwinship  Is there a reason why we have a check on daddr, port and proto for individual service chan mark-for-masquerade jump rule? 
Packets will only be jumped to this chain when they match on daddr, port and proto. 
``` 
chain service-JKB34KST-default/test/tcp/http {
	ip daddr 10.96.83.93 tcp dport 80 ip saddr != 10.244.0.0/16 jump mark-for-masquerade
	numgen random ...
}

# maybe this can be simplified to

chain service-JKB34KST-default/test/tcp/http {
	ip saddr != 10.244.0.0/16 jump mark-for-masquerade
	numgen random ...
}

```

**Comment by performantdata** [comment]
I don't see `nftables` mentioned as a `ProxyMode` option in [the documentation](https://kubernetes.io/docs/reference/config-api/kube-proxy-config.v1alpha1/#kubeproxy-config-k8s-io-v1alpha1-ProxyMode).

ref.: https://kubernetes.slack.com/archives/C09QYUH5W/p1733790390408679

**Comment by danwinship** [comment]
that was _just_ fixed yesterday :slightly_smiling_face: #128698

**Comment by danwinship** [comment]
This probably isn't needed any more
